File: graphs/knightMinimumSteps.js
/**
 * Given a square chessboard of N x N size, find the minimum number of steps a Knight will take
 * to reach from a given starting position to a target position. The positions are given as
 * (x, y) coordinates on the chessboard.
 */
function knightMinimumSteps(chessboardSize, startPos, targetPos) {
  // All possible moves a knight can make
  const moves = [
    [-2, -1], [-2, 1], [-1, -2], [-1, 2],
    [1, -2], [1, 2], [2, -1], [2, 1]
  ];

  // Create a queue for BFS with initial position and steps
  const queue = [[startPos[0], startPos[1], 0]];

  // Keep track of visited positions
  const visited = new Set();
  visited.add(`${startPos[0]},${startPos[1]}`);

  while (queue.length > 0) {
    const [x, y, steps] = queue.shift();

    // If we've reached the target, return steps
    if (x === targetPos[0] && y === targetPos[1]) {
      return steps;
    }

    // Try all possible knight moves
    for (const [dx, dy] of moves) {
      const newX = x + dx;
      const newY = y + dy;

      // Check if the new position is valid and unvisited
      if (isValidPosition(newX, newY, chessboardSize) &&
          !visited.has(`${newX},${newY}`)) {
        queue.push([newX, newY, steps + 1]);
        visited.add(`${newX},${newY}`);
      }
    }
  }

  // If target is unreachable
  return -1;
}

function isValidPosition(x, y, size) {
  return x >= 0 && x < size && y >= 0 && y < size;
}

console.log(knightMinimumSteps(8, [0, 0], [7, 7]));

module.exports = {
  knightMinimumSteps,
}

File: graphs/python/knight_minimum_steps.py
"""
Functions for finding minimum steps for a knight to reach a target position on a chessboard.
"""
from collections import deque


def add_coordinates(first, second):
    return (sum(pair) for pair in zip(first, second))


def in_bounds(size, pos):
    x, y = pos

    return 0 <= x < size and 0 <= y < size


def knight_minimum_steps(chessboard_size, start_pos, target_pos):
    moves = [
        (-2, -1), (-2, 1), (-1, -2), (-1, 2),
        (1, -2), (1, 2), (2, -1), (2, 1)
    ]

    queue = deque([(start_pos, 0)])

    visited = {start_pos}

    while queue:
        pos, steps = queue.popleft()

        if pos == target_pos:
            return steps

        for move in moves:
            new_pos = add_coordinates(pos, move)

            if new_pos in visited or not in_bounds(chessboard_size, new_pos):
                continue

            queue.append((new_pos, steps + 1))
            visited.add(new_pos)

    return -1


if __name__ == '__main__':
    print(knight_minimum_steps(8, (3, 3), (4, 5)))
    print(knight_minimum_steps(8, (3, 3), (4, 4)))
    print(knight_minimum_steps(8, (3, 3), (3, 4)))
    print(knight_minimum_steps(8, (3, 3), (5, 5)))
    print(knight_minimum_steps(8, (0, 0), (7, 7)))
